cmd/edgectl/aes_install.go

```diff
@@ -0,0 +1,290 @@
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/datawire/ambassador/pkg/k8s"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+func aesInstallCmd() *cobra.Command {
+	res := &cobra.Command{
+		Use:   "install",
+		Short: "Install the Ambassador Edge Stack in your cluster",
+		Args:  cobra.ExactArgs(0),
+		RunE:  aesInstall,
+	}
+	_ = res.Flags().StringP(
+		"context", "c", "",
+		"The Kubernetes context to use. Defaults to the current kubectl context.",
+	)
+	_ = res.Flags().StringP(
+		"namespace", "n", "",
+		"The Kubernetes namespace to use. Defaults to kubectl's default for the context.",
+	)
+
+	return res
+}
+
+func aesInstall(cmd *cobra.Command, args []string) error {
+	metrics := NewMetrics()
+	_ = metrics.Report("install")
+
+	// Display version information
+	fmt.Printf("-> Installing the Ambassador Edge Stack %s\n", Version)
+
+	// Attempt to talk to the specified cluster
+	context, _ := cmd.Flags().GetString("context")
+	namespace, _ := cmd.Flags().GetString("namespace")
+	kubeinfo := k8s.NewKubeInfo("", context, namespace)
+	i := &Installer{
+		kubeinfo,
+	}
+	if err := i.ShowKubectl("cluster-info", "cluster-info"); err != nil {
+		return err
+	}
+
+	if err := i.ShowKubectl("install CRDs", "apply", "-f", "https://www.getambassador.io/yaml/aes-crds.yaml"); err != nil {
+		return err
+	}
+
+	if err := i.ShowKubectl("wait for CRDs", "wait", "--for", "condition=established", "--timeout=90s", "crd", "-lproduct=aes"); err != nil {
+		return err
+	}
+
+	if err := i.ShowKubectl("install AES", "apply", "-f", "https://www.getambassador.io/yaml/aes.yaml"); err != nil {
+		return err
+	}
+
+	if err := i.ShowKubectl("wait for AES", "-n", "ambassador", "wait", "--for", "condition=available", "--timeout=90s", "deploy", "-lproduct=aes"); err != nil {
+		return err
+	}
+
+	_ = metrics.Report("deploy") // TODO: Send cluster type and Helm version
+
+	ipAddress := ""
+	for {
+		var err error
+		ipAddress, err = i.CaptureKubectl("get IP address", "get", "-n", "ambassador", "service", "ambassador", "-o", `go-template={{range .status.loadBalancer.ingress}}{{print .ip "\n"}}{{end}}`)
+		if err != nil {
+			return err
+		}
+		ipAddress = strings.TrimSpace(ipAddress)
+		if ipAddress != "" {
+			break
+		}
+		time.Sleep(250 * time.Millisecond) // FIXME: Time out at some point...
+	}
+
+	fmt.Println("\nYour IP address is", ipAddress)
+
+	// Wait for Ambassador to be ready to serve ACME requests.
+	// FIXME: This assumes we can connect to the load balancer. If this
+	// assumption is incorrect, this code will loop forever.
+	for {
+		// FIXME: Time out at some point...
+		time.Sleep(500 * time.Millisecond)
+		resp, err := http.Get("http://" + ipAddress + "/.well-known/acme-challenge/")
+		if err != nil {
+			fmt.Printf("Waiting for Ambassador (get): %#v\n", err)
+			continue
+		}
+		_, _ = ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		if resp.StatusCode != 404 {
```

Comment:
I would not perform these StatusCode and Header checks on the client-side. If we can open a socket and get any kind of response other than the handled `err`, we should be good to proceed.

Reply:
Isn't `upstream request timeout` coming from Envoy? That'd be `err == nil` and `StatusCode == 503` (or whatever), right?

Reply:
I think you right, that would make much more sense!